File: backend/routes/weak_topics.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, model_validator
from typing import List, Any
import json

from backend.lib.clients.groq import chat_with_context
from backend.lib.prompts.templates import WEAK_TOPICS_SYSTEM_PROMPT

router = APIRouter()


class IncorrectQuestion(BaseModel):
    question: str
    user_answer: str = ""
    correct_answer: str = ""

    @model_validator(mode="before")
    @classmethod
    def handle_frontend_fields(cls, data: Any) -> Any:
        """Accept 'selected' and 'correct' field names sent by QuizPane.jsx."""
        if isinstance(data, dict):
            if "selected" in data and "user_answer" not in data:
                data["user_answer"] = data["selected"]
            if "correct" in data and "correct_answer" not in data:
                data["correct_answer"] = data["correct"]
        return data


class WeakTopicsRequest(BaseModel):
    subject: str
    incorrect_questions: List[IncorrectQuestion]


@router.post("/weak-topics")
async def analyze_weaknesses(req: WeakTopicsRequest):
    if not req.incorrect_questions:
        return []

    questions_context = ""
    for i, q in enumerate(req.incorrect_questions):
        questions_context += (
            f"Q{i+1}: {q.question}\n"
            f"User selected: {q.user_answer}\n"
            f"Correct answer: {q.correct_answer}\n\n"
        )

    sys_prompt = WEAK_TOPICS_SYSTEM_PROMPT

    messages = [
        {"role": "system", "content": sys_prompt},
        {"role": "user", "content": f"Subject: {req.subject}\n\nMissed Questions:\n{questions_context}"},
    ]

    try:
        response_text = chat_with_context(messages)
    except Exception as e:
        print(f"[ERROR] Groq call failed in weak-topics: {e}")
        raise HTTPException(status_code=502, detail="AI service temporarily unavailable.")

    if not response_text or response_text.startswith("Error:"):
        raise HTTPException(status_code=502, detail=f"AI service error: {response_text}")

    try:
        cleaned = response_text.replace("```json", "").replace("```", "").strip()
        analysis = json.loads(cleaned)
        if not isinstance(analysis, list):
            raise ValueError("LLM did not return a JSON list")
        return analysis
    except Exception as e:
        print(f"Error parsing weak topics JSON: {e}\nRaw: {response_text}")
        raise HTTPException(status_code=400, detail="Invalid JSON from LLM")

File: api/weak-topics.ts
import type { VercelRequest, VercelResponse } from '@vercel/node';
import { chatWithContext } from '../lib/groq';
import { WEAK_TOPICS_SYSTEM_PROMPT } from '../lib/prompts';

interface IncorrectQuestion {
  question: string;
  user_answer?: string;
  correct_answer?: string;
  selected?: string;
  correct?: string;
}

interface WeakTopicsRequest {
  subject: string;
  incorrect_questions: IncorrectQuestion[];
}

export default async function handler(
  req: VercelRequest,
  res: VercelResponse
) {
  if (req.method !== 'POST') {
    return res.status(405).json({ detail: 'Method Not Allowed' });
  }

  const body = req.body as WeakTopicsRequest;
  if (!body || !body.incorrect_questions || body.incorrect_questions.length === 0) {
    return res.status(200).json([]);
  }

  let questions_context = "";
  body.incorrect_questions.forEach((q, i) => {
    // Handle frontend fields mapping identical to Python's handle_frontend_fields
    const user_answer = q.user_answer || q.selected || "";
    const correct_answer = q.correct_answer || q.correct || "";

    questions_context += `Q${i + 1}: ${q.question}\nUser selected: ${user_answer}\nCorrect answer: ${correct_answer}\n\n`;
  });

  const messages = [
    { role: "system", content: WEAK_TOPICS_SYSTEM_PROMPT },
    { role: "user", content: `Subject: ${body.subject}\n\nMissed Questions:\n${questions_context}` },
  ];

  let response_text = "";
  try {
    response_text = await chatWithContext(messages, { temperature: 0.7 });
  } catch (error: any) {
    console.error(`[ERROR] Groq call failed in weak-topics: ${error}`);
    return res.status(502).json({ detail: "AI service temporarily unavailable." });
  }

  if (!response_text || response_text.startsWith("Error:")) {
    return res.status(502).json({ detail: `AI service error: ${response_text}` });
  }

  try {
    const cleaned = response_text.replace(/```json/g, "").replace(/```/g, "").trim();
    const analysis = JSON.parse(cleaned);
    
    if (!Array.isArray(analysis)) {
      throw new Error("LLM did not return a JSON list");
    }
    
    return res.status(200).json(analysis);
  } catch (error: any) {
    console.error(`Error parsing weak topics JSON: ${error}\nRaw: ${response_text}`);
    return res.status(400).json({ detail: "Invalid JSON from LLM" });
  }
}
